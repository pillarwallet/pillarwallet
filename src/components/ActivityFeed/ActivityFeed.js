// @flow
/*
    Pillar Wallet: the personal data locker
    Copyright (C) 2019 Stiftung Pillar Project

    This program is free software; you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation; either version 2 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License along
    with this program; if not, write to the Free Software Foundation, Inc.,
    51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
*/
import * as React from 'react';
import { FlatList } from 'react-native';
import isEqual from 'lodash.isequal';
import type { NavigationScreenProp } from 'react-navigation';
import styled, { withTheme } from 'styled-components/native';
import memoize from 'memoize-one';

// types
import type { EnsRegistry } from 'reducers/ensRegistryReducer';

// components
import Title from 'components/Title';
import Tabs from 'components/Tabs';
import { BaseText, MediumText } from 'components/Typography';
import EmptyStateParagraph from 'components/EmptyState/EmptyStateParagraph';
import ActivityFeedItem from 'components/ActivityFeed/ActivityFeedItem';
import EventDetails, { shouldShowEventDetails } from 'components/EventDetails';
import Modal from 'components/Modal';

// utils
import { groupAndSortByDate } from 'utils/common';
import { fontStyles, spacing } from 'utils/variables';

// constants
import { COLLECTIBLE_TRANSACTION } from 'constants/collectiblesConstants';
import { TRANSACTION_EVENT } from 'constants/historyConstants';
import { USER_EVENT } from 'constants/userEventsConstants';
import { BADGE_REWARD_EVENT } from 'constants/badgesConstants';
import { LIGHT_THEME } from 'constants/appSettingsConstants';
import { SABLIER_EVENT } from 'constants/sablierConstants';


const ActivityFeedList = styled.FlatList`
  width: 100%;
  flex: 1;
`;

const ActivityFeedWrapper = styled.View`
  flex: 1;
`;

const ActivityFeedHeader = styled.View`
  padding: ${spacing.mediumLarge}px ${spacing.large}px 0;
`;

const SectionHeaderWrapper = styled.View`
  width: 100%;
  padding: ${spacing.small}px ${spacing.large}px;
`;

const SectionHeader = styled(BaseText)`
  ${fontStyles.regular};
  color: ${({ theme }) => theme.colors.basic010};
`;

const EmptyStateWrapper = styled.View`
  padding: 15px 30px 30px;
  width: 100%;
  align-items: center;
  justify-content: center;
`;

const CardHeaderWrapper = styled.View`
  border-top-left-radius: 30px;
  border-top-right-radius: 30px;
  background-color: ${({ theme }) => theme.colors.basic050};
  padding: 13px 20px 17px;
  margin-top: 16px;
  ${({ theme }) => theme.current === LIGHT_THEME && `
    box-shadow: 0px 2px 7px rgba(0,0,0,.1);
    elevation: 1;
  `}
`;

const CardBackgroundWrapper = styled.View`
  ${({ theme, card }) => card && `background-color: ${theme.colors.basic050};`}
`;

type EmptyState = {|
  title?: string,
  textBody?: string,
|};

type Tab = {|
  id: string,
  name: string,
  icon?: string,
  onPress: Function,
  unread?: number,
  tabImageNormal?: string,
  tabImageActive?: string,
  data: Object[],
  emptyState?: EmptyState,
|};

type Props = {
  navigation: NavigationScreenProp<*>,
  feedTitle?: string,
  wrapperStyle?: Object,
  contentContainerStyle?: Object,
  initialNumToRender: number,
  tabs?: Tab[],
  activeTab?: string,
  feedData?: Object[],
  extraFeedData?: Object[],
  hideTabs: boolean,
  emptyState?: EmptyState,
  ensRegistry: EnsRegistry,
  tabsComponent?: React.Node,
  headerComponent?: React.Node,
  flatListProps?: FlatList,
  isPPNView?: boolean,
  isForAllAccounts?: boolean,
  isAssetView?: boolean,
  card?: boolean,
  cardHeaderTitle?: string,
};

const ITEM_TYPE = {
  HEADER: 'HEADER',
  TABS: 'TABS',
  SECTION: 'SECTION',
  ITEM: 'ITEM',
  EMPTY_STATE: 'EMPTY_STATE',
  CARD_HEADER: 'CARD_HEADER',
};

class ActivityFeed extends React.Component<Props> {
  static defaultProps = {
    initialNumToRender: 7,
  };

  generateFeedSections = memoize(
    (tabs, activeTab, feedData, headerComponent, tabsComponent, card) => {
      let feedList = feedData || [];

      if (tabs.length) {
        const activeTabInfo = tabs.find(({ id }) => id === activeTab);
        if (activeTabInfo) ({ data: feedList } = activeTabInfo);
      }

      const filteredFeedList = feedList.filter(this.shouldRenderActivityItem);

      const dataSections = groupAndSortByDate(filteredFeedList);

      const items = [];
      items.push({ type: ITEM_TYPE.HEADER, component: headerComponent });
      items.push({ type: ITEM_TYPE.TABS, component: tabsComponent });
      if (!filteredFeedList.length) {
        items.push({ type: ITEM_TYPE.EMPTY_STATE });
      } else {
        if (card) {
          items.push({ type: ITEM_TYPE.CARD_HEADER });
        }
        dataSections.forEach(({ data, ...section }) => {
          items.push({ type: ITEM_TYPE.SECTION, section });
          data.forEach(item => items.push({ type: ITEM_TYPE.ITEM, item }));
        });
      }

      return items;
    },
    isEqual,
  );

  getEmptyStateData = () => {
    const {
      tabs = [],
      activeTab,
      emptyState,
    } = this.props;

    let emptyStateData = emptyState || {};
    if (tabs.length) {
      const activeTabInfo = tabs.find(({ id }) => id === activeTab);
      if (activeTabInfo) ({ emptyState: emptyStateData = {} } = activeTabInfo);
    }
    return emptyStateData;
  }

  shouldComponentUpdate(nextProps: Props) {
    return !isEqual(this.props, nextProps);
  }

  selectEvent = (eventData: Object, itemData: Object) => {
    if (!shouldShowEventDetails(eventData)) return;

    Modal.open(() => (
      <EventDetails
        event={eventData}
        itemData={itemData}
        navigation={this.props.navigation}
        isForAllAccounts={this.props.isForAllAccounts}
      />
    ));
  };

  shouldRenderActivityItem = (item: Object) => {
    const typesThatRender = [
      USER_EVENT, TRANSACTION_EVENT, COLLECTIBLE_TRANSACTION, BADGE_REWARD_EVENT, SABLIER_EVENT,
    ];
    return typesThatRender.includes(item.type);
  };

  renderActivityFeedItem = ({ item }) => {
    const {
      isPPNView,
      isForAllAccounts,
      isAssetView,
      card,
      cardHeaderTitle,
    } = this.props;
    switch (item.type) {
      case ITEM_TYPE.HEADER:
      case ITEM_TYPE.TABS:
        return item.component;
      case ITEM_TYPE.EMPTY_STATE:
        const emptyStateData = this.getEmptyStateData();
        return (
          <EmptyStateWrapper>
            <EmptyStateParagraph {...emptyStateData} />
          </EmptyStateWrapper>
        );
      case ITEM_TYPE.SECTION:
        return (
          <CardBackgroundWrapper card={card}>
            <SectionHeaderWrapper>
              <SectionHeader>{item.section.title}</SectionHeader>
            </SectionHeaderWrapper>
          </CardBackgroundWrapper>
        );
      case ITEM_TYPE.CARD_HEADER:
        return (
          <CardHeaderWrapper>
            <MediumText big>{cardHeaderTitle}</MediumText>
          </CardHeaderWrapper>
        );
      default:
        return (
          <CardBackgroundWrapper card={card}>
            <ActivityFeedItem
              event={item.item}
              selectEvent={this.selectEvent}
              isPPNView={isPPNView}
              isAssetView={isAssetView}
              isForAllAccounts={isForAllAccounts}
            />
          </CardBackgroundWrapper>
        );
    }
  };

  getActivityFeedListKeyExtractor = (item: Object = {}) => {
    switch (item.type) {
      case ITEM_TYPE.HEADER:
      case ITEM_TYPE.TABS:
      case ITEM_TYPE.EMPTY_STATE:
      case ITEM_TYPE.CARD_HEADER:
        return item.type;
      case ITEM_TYPE.SECTION:
        return item.section.title;
      default:
        const { createdAt = '' } = item.item;
        return `${createdAt.toString()}${item.item.id || item.item._id || item.item.hash || ''}${item.item.from || ''}`;
    }
  };

  render() {
    const {
      feedTitle,
      wrapperStyle,
      contentContainerStyle,
      initialNumToRender,
      tabs = [],
      activeTab,
      extraFeedData,
      hideTabs,
      feedData,
      headerComponent,
      tabsComponent,
      flatListProps,
      card,
    } = this.props;

    const formattedFeedData = this.generateFeedSections(
      tabs, activeTab, feedData, headerComponent, tabsComponent, card,
    );

    const firstTab = tabs.length ? tabs[0].id : '';

    const additionalContentContainerStyle = !formattedFeedData.length
      ? { justifyContent: 'center', flex: 1 }
      : {};

    const tabsProps = tabs.map(({ data, emptyState, ...necessaryTabProps }) => necessaryTabProps);

    return (
      <ActivityFeedWrapper style={wrapperStyle}>
        {!!feedTitle &&
        <ActivityFeedHeader>
          <Title subtitle title={feedTitle} noMargin />
        </ActivityFeedHeader>}
        {tabs.length > 1 && !hideTabs &&
          <Tabs
            tabs={tabsProps}
            wrapperStyle={{ paddingTop: 0 }}
            activeTab={activeTab || firstTab}
          />
        }
        <ActivityFeedList
          data={formattedFeedData}
          initialNumToRender={initialNumToRender}
          extraData={extraFeedData}
          renderItem={this.renderActivityFeedItem}
          onEndReachedThreshold={0.5}
          keyExtractor={this.getActivityFeedListKeyExtractor}
          contentContainerStyle={[additionalContentContainerStyle, contentContainerStyle]}
          stickyHeaderIndices={[1]}
          {...flatListProps}
        />
      </ActivityFeedWrapper>
    );
  }
}

export default withTheme(ActivityFeed);
