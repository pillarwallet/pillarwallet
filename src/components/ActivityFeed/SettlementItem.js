// @flow
/*
    Pillar Wallet: the personal data locker
    Copyright (C) 2019 Stiftung Pillar Project

    This program is free software; you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation; either version 2 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License along
    with this program; if not, write to the Free Software Foundation, Inc.,
    51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
*/
import * as React from 'react';
import styled from 'styled-components/native';
import isEmpty from 'lodash.isempty';
import BigNumber from 'bignumber.js';

import ListItemWithImage from 'components/ListItem/ListItemWithImage';
import { BaseText } from 'components/Typography';
import TankAssetBalance from 'components/TankAssetBalance';
import { fontSizes, spacing } from 'utils/variables';
import { themedColors } from 'utils/themes';
import { SYNTHETIC, NONSYNTHETIC } from 'constants/assetsConstants';
import { getAssetData } from 'utils/assets';
import { formatAmount, formatUnits } from 'utils/common';
import type { Asset } from 'models/Asset';

const ppnIcon = require('assets/icons/icon_PPN.png');

type SettlementItemType = {
  hash: string,
  symbol: string,
  value: string | number,
}

type Props = {
  settleData: SettlementItemType[],
  type?: string,
  asset?: string,
  onPress: Function,
  isPending?: boolean,
  accountAssets: Asset[],
  supportedAssets: Asset[],
}

const ListWrapper = styled.View`
  align-items: flex-end;
  padding-left: ${spacing.mediumLarge}px;
`;

const ItemValue = styled(BaseText)`
  font-size: ${fontSizes.big}px;
  color: ${themedColors.primary};
  text-align: right;
`;

export const SettlementItem = (props: Props) => {
  const {
    settleData,
    type,
    asset,
    onPress,
    isPending,
    accountAssets,
    supportedAssets,
  } = props;
  const ppnTransactions = asset
    ? settleData.filter(({ symbol }) => symbol === asset)
    : settleData;

  const valueByAsset: Object = {};
  const formattedPPNTransactions = ppnTransactions.map((trx) => {
    const { symbol, value: rawValue } = trx;
    const { decimals = 18 } = getAssetData(accountAssets, supportedAssets, symbol);
    const value = new BigNumber(rawValue);
    if (!valueByAsset[symbol]) {
      valueByAsset[symbol] = { ...trx, value, decimals };
    } else {
      const { value: currentValue } = valueByAsset[symbol];
      valueByAsset[symbol].value = currentValue.plus(value);
    }
    const formatted = formatAmount(formatUnits(value.toString(), decimals));
    return { ...trx, formatted };
  });

  const valuesArray = Object.keys(valueByAsset).map((key) => valueByAsset[key]);

  const itemStatusIcon = (isPending && 'pending') || '';
  const rightColumnInnerStyle = (isPending && { flexDirection: 'row', alignItems: 'center' }) || {};
  const customAddonAlignLeft = !isEmpty(rightColumnInnerStyle);

  return (
    <React.Fragment>
      {(!type || type === NONSYNTHETIC) &&
        <ListItemWithImage
          onPress={onPress}
          label="Deposit"
          itemImageSource={ppnIcon}
          subtext="to Smart Wallet"
          customAddon={(
            <ListWrapper>
              {valuesArray.map(({ symbol, value, decimals }) => {
                const formatted = formatAmount(formatUnits(value.toString(), decimals));
                return (<ItemValue key={symbol}>{`${formatted} ${symbol}`}</ItemValue>);
              })}
            </ListWrapper>)}
          innerWrapperHorizontalAlign="flex-start"
          itemStatusIcon={itemStatusIcon}
          customAddonAlignLeft={customAddonAlignLeft}
          rightColumnInnerStyle={rightColumnInnerStyle}
          diameter={56}
        />
      }
      {(!type || type === SYNTHETIC) &&
        <ListItemWithImage
          onPress={onPress}
          label="Withdrawal"
          itemImageSource={ppnIcon}
          subtext="from PLR Network"
          customAddon={(
            <ListWrapper>
              {formattedPPNTransactions.map(({ symbol, formatted, hash }) => (
                <TankAssetBalance
                  key={hash}
                  amount={`- ${formatted} ${symbol}`}
                  monoColor
                />
              ))}
            </ListWrapper>)}
          innerWrapperHorizontalAlign="flex-start"
          itemStatusIcon={itemStatusIcon}
          customAddonAlignLeft={customAddonAlignLeft}
          rightColumnInnerStyle={rightColumnInnerStyle}
          diameter={56}
        />
      }
    </React.Fragment>
  );
};
