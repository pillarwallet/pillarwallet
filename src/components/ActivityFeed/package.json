{
  "name": "ActivityFeed.js",
  "version": "0.0.0",
  "private": true,
  "main": "./ActivityFeed.js"
}
