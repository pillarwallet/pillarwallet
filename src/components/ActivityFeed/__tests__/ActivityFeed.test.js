// @flow
/*
    Pillar Wallet: the personal data locker
    Copyright (C) 2019 Stiftung Pillar Project

    This program is free software; you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation; either version 2 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License along
    with this program; if not, write to the Free Software Foundation, Inc.,
    51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
*/
import * as React from 'react';
import { Provider } from 'react-redux';
import renderer from 'react-test-renderer';
import thunk from 'redux-thunk';
import configureMockStore from 'redux-mock-store';
import { ThemeProvider } from 'styled-components';
import { defaultTheme } from 'utils/themes';
import ActivityFeed from 'components/ActivityFeed/ActivityFeed';
import { I18nextProvider } from 'react-i18next';
import i18n from 'services/localisation/testing';
import { TRANSACTION_EVENT } from 'constants/historyConstants';

import { initialState as historyState } from 'reducers/historyReducer';
import { initialState as assetsState } from 'reducers/assetsReducer';
import { initialState as accountsState } from 'reducers/accountsReducer';
import { initialState as ensRegistryState } from 'reducers/ensRegistryReducer';
import { initialState as smartWalletState } from 'reducers/smartWalletReducer';
import { initialState as referralsState } from 'reducers/referralsReducer';
import { initialState as sessionState } from 'reducers/sessionReducer';


const mockStore = configureMockStore([thunk]);

const initialStore = mockStore({
  assets: assetsState,
  accounts: accountsState,
  history: historyState,
  ensRegistry: ensRegistryState,
  smartWallet: smartWalletState,
  referrals: referralsState,
  session: sessionState,
});

const Component = (store, children) => (
  renderer.create(
    <Provider store={store}>
      <I18nextProvider i18n={i18n}>
        {children}
      </I18nextProvider>
    </Provider>)
);

const transaction = (attrs = {}) => {
  return {
    type: TRANSACTION_EVENT,
    createdAt: '2019-09-09 12:00',
    value: '0',
    from: '0x0000000000000000000000000000000000000000',
    to: '0x0000000000000000000000000000000000000000',
    ...attrs,
  };
};

describe('ActivityFeed', () => {
  it('renders the Asset correctly', () => {
    const transactions = [
      transaction(),
    ];

    const tabs = [{ data: transactions }];

    const component = Component(initialStore,
      <ThemeProvider theme={defaultTheme}>
        <ActivityFeed tabs={tabs} />
      </ThemeProvider>).toJSON();

    expect(component).toMatchSnapshot();
  });

  it('does not fail for invalid values', () => {
    const transactions = [
      transaction({ value: undefined }),
    ];

    const tabs = [{ data: transactions }];

    const component = Component(initialStore,
      <ThemeProvider theme={defaultTheme}>
        <ActivityFeed tabs={tabs} />
      </ThemeProvider>).toJSON();

    expect(component).toMatchSnapshot();
  });
});
