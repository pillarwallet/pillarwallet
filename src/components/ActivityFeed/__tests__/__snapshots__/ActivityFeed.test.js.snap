// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ActivityFeed does not fail for invalid values 1`] = `
<View
  style={
    Array [
      Object {
        "flexBasis": 0,
        "flexGrow": 1,
        "flexShrink": 1,
      },
      undefined,
    ]
  }
>
  <RCTScrollView
    contentContainerStyle={
      Array [
        Object {},
        undefined,
      ]
    }
    data={
      Array [
        Object {
          "component": undefined,
          "type": "HEADER",
        },
        Object {
          "component": undefined,
          "type": "TABS",
        },
        Object {
          "section": Object {
            "date": "Invalid Date",
            "title": "Invalid Date",
          },
          "type": "SECTION",
        },
        Object {
          "item": Object {
            "createdAt": "2019-09-09 12:00",
            "from": "0x0000000000000000000000000000000000000000",
            "to": "0x0000000000000000000000000000000000000000",
            "type": "transactionEvent",
            "value": undefined,
          },
          "type": "ITEM",
        },
      ]
    }
    disableVirtualization={false}
    getItem={[Function]}
    getItemCount={[Function]}
    horizontal={false}
    initialNumToRender={7}
    keyExtractor={[Function]}
    maxToRenderPerBatch={10}
    numColumns={1}
    onContentSizeChange={[Function]}
    onEndReachedThreshold={0.5}
    onLayout={[Function]}
    onMomentumScrollEnd={[Function]}
    onScroll={[Function]}
    onScrollBeginDrag={[Function]}
    onScrollEndDrag={[Function]}
    removeClippedSubviews={false}
    renderItem={[Function]}
    scrollEventThrottle={50}
    stickyHeaderIndices={
      Array [
        1,
      ]
    }
    style={
      Array [
        Object {
          "flexBasis": 0,
          "flexGrow": 1,
          "flexShrink": 1,
          "width": "100%",
        },
        undefined,
      ]
    }
    updateCellsBatchingPeriod={50}
    viewabilityConfigCallbackPairs={Array []}
    windowSize={21}
  >
    <View>
      <View
        onLayout={[Function]}
        style={null}
      />
      <View
        onLayout={[Function]}
        style={null}
      />
      <View
        onLayout={[Function]}
        style={null}
      >
        <View
          style={
            Array [
              Object {
                "paddingBottom": 8,
                "paddingLeft": 20,
                "paddingRight": 20,
                "paddingTop": 8,
                "width": "100%",
              },
              undefined,
            ]
          }
        >
          <Text
            style={
              Array [
                Object {
                  "color": "#0A1427",
                  "fontFamily": "EuclidCircularB-Regular",
                  "textAlignVertical": "center",
                },
                Array [
                  Object {
                    "color": "#8B939E",
                    "fontSize": 14,
                    "lineHeight": 22,
                  },
                  undefined,
                ],
              ]
            }
          >
            Invalid Date
          </Text>
        </View>
      </View>
      <View
        onLayout={[Function]}
        style={null}
      >
        <View
          style={
            Array [
              Object {
                "flexDirection": "column",
                "width": "100%",
              },
              undefined,
            ]
          }
        >
          <View
            accessible={true}
            focusable={true}
            isTVSelectable={true}
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              Object {
                "alignItems": "center",
                "flexDirection": "row",
                "justifyContent": "center",
                "opacity": 1,
                "paddingBottom": 14,
                "paddingLeft": 20,
                "paddingRight": 20,
                "paddingTop": 14,
                "width": "100%",
              }
            }
          >
            <View
              style={
                Array [
                  Object {
                    "alignItems": "center",
                    "justifyContent": "center",
                    "paddingRight": 12,
                  },
                  undefined,
                ]
              }
            >
              <View
                backgroundColor={null}
                diameter={48}
                style={
                  Array [
                    Object {
                      "alignItems": "center",
                      "backgroundColor": "#EBF0F6",
                      "borderRadius": 24,
                      "height": 48,
                      "justifyContent": "center",
                      "overflow": "hidden",
                      "textAlign": "center",
                      "width": 48,
                    },
                    undefined,
                  ]
                }
              >
                <View
                  name="sent"
                  style={
                    Array [
                      Object {
                        "color": "#BD573A",
                        "fontSize": 48,
                      },
                      undefined,
                    ]
                  }
                />
              </View>
            </View>
            <View
              style={
                Object {
                  "flex": 1,
                }
              }
            >
              <View
                style={
                  Array [
                    Object {
                      "alignItems": "center",
                      "flexDirection": "row",
                      "justifyContent": "space-between",
                      "width": "100%",
                    },
                    undefined,
                  ]
                }
                type="ACTION"
              >
                <View
                  style={
                    Array [
                      Object {
                        "alignItems": "flex-start",
                        "flexBasis": 0,
                        "flexDirection": "column",
                        "flexGrow": 1,
                        "flexShrink": 1,
                        "justifyContent": "center",
                        "marginTop": 0,
                        "minHeight": 54,
                      },
                      Object {
                        "flexGrow": 1,
                      },
                    ]
                  }
                  type="ACTION"
                >
                  <View
                    style={
                      Array [
                        Object {
                          "alignItems": "center",
                          "flexDirection": "row",
                          "justifyContent": "space-between",
                        },
                        undefined,
                      ]
                    }
                  >
                    <Text
                      ellipsizeMode="tail"
                      numberOfLines={2}
                      style={
                        Array [
                          Object {
                            "color": "#0A1427",
                            "fontFamily": "EuclidCircularB-Regular",
                            "textAlignVertical": "center",
                          },
                          Array [
                            Object {
                              "color": "#0A1427",
                              "fontFamily": "EuclidCircularB-Medium",
                              "textAlignVertical": "center",
                            },
                            Array [
                              Object {
                                "color": "#0A1427",
                                "fontSize": 18,
                                "lineHeight": 28,
                                "width": "100%",
                              },
                              undefined,
                            ],
                          ],
                        ]
                      }
                      type="ACTION"
                    >
                      0x0000â€¦000000
                    </Text>
                  </View>
                </View>
                <View
                  rightColumn={true}
                  style={
                    Array [
                      Object {
                        "alignItems": "flex-end",
                        "flexDirection": "column",
                        "justifyContent": "center",
                        "marginLeft": 10,
                        "marginTop": 0,
                        "minHeight": 54,
                      },
                      Object {
                        "maxWidth": "50%",
                      },
                    ]
                  }
                  type="ACTION"
                >
                  <View
                    style={
                      Array [
                        undefined,
                        Object {
                          "flexWrap": "wrap",
                        },
                      ]
                    }
                  >
                    <View
                      horizontal={true}
                      style={
                        Array [
                          Object {
                            "flexDirection": "row",
                            "marginBottom": 0,
                            "marginLeft": 0,
                            "marginRight": 0,
                            "marginTop": 0,
                          },
                          Object {
                            "alignItems": "center",
                            "flexWrap": "wrap",
                            "justifyContent": "flex-end",
                          },
                        ]
                      }
                    >
                      <Text
                        color="#0A1427"
                        ellipsizeMode="tail"
                        numberOfLines={2}
                        style={
                          Array [
                            Object {
                              "color": "#0A1427",
                              "fontFamily": "EuclidCircularB-Regular",
                              "textAlignVertical": "center",
                            },
                            Array [
                              Object {
                                "color": "#0A1427",
                                "fontSize": 18,
                                "lineHeight": 28,
                                "textAlign": "right",
                              },
                              undefined,
                            ],
                          ]
                        }
                      >
                         0 undefined
                      </Text>
                    </View>
                  </View>
                </View>
              </View>
            </View>
          </View>
        </View>
      </View>
    </View>
  </RCTScrollView>
</View>
`;

exports[`ActivityFeed renders the Asset correctly 1`] = `
<View
  style={
    Array [
      Object {
        "flexBasis": 0,
        "flexGrow": 1,
        "flexShrink": 1,
      },
      undefined,
    ]
  }
>
  <RCTScrollView
    contentContainerStyle={
      Array [
        Object {},
        undefined,
      ]
    }
    data={
      Array [
        Object {
          "component": undefined,
          "type": "HEADER",
        },
        Object {
          "component": undefined,
          "type": "TABS",
        },
        Object {
          "section": Object {
            "date": "Invalid Date",
            "title": "Invalid Date",
          },
          "type": "SECTION",
        },
        Object {
          "item": Object {
            "createdAt": "2019-09-09 12:00",
            "from": "0x0000000000000000000000000000000000000000",
            "to": "0x0000000000000000000000000000000000000000",
            "type": "transactionEvent",
            "value": "0",
          },
          "type": "ITEM",
        },
      ]
    }
    disableVirtualization={false}
    getItem={[Function]}
    getItemCount={[Function]}
    horizontal={false}
    initialNumToRender={7}
    keyExtractor={[Function]}
    maxToRenderPerBatch={10}
    numColumns={1}
    onContentSizeChange={[Function]}
    onEndReachedThreshold={0.5}
    onLayout={[Function]}
    onMomentumScrollEnd={[Function]}
    onScroll={[Function]}
    onScrollBeginDrag={[Function]}
    onScrollEndDrag={[Function]}
    removeClippedSubviews={false}
    renderItem={[Function]}
    scrollEventThrottle={50}
    stickyHeaderIndices={
      Array [
        1,
      ]
    }
    style={
      Array [
        Object {
          "flexBasis": 0,
          "flexGrow": 1,
          "flexShrink": 1,
          "width": "100%",
        },
        undefined,
      ]
    }
    updateCellsBatchingPeriod={50}
    viewabilityConfigCallbackPairs={Array []}
    windowSize={21}
  >
    <View>
      <View
        onLayout={[Function]}
        style={null}
      />
      <View
        onLayout={[Function]}
        style={null}
      />
      <View
        onLayout={[Function]}
        style={null}
      >
        <View
          style={
            Array [
              Object {
                "paddingBottom": 8,
                "paddingLeft": 20,
                "paddingRight": 20,
                "paddingTop": 8,
                "width": "100%",
              },
              undefined,
            ]
          }
        >
          <Text
            style={
              Array [
                Object {
                  "color": "#0A1427",
                  "fontFamily": "EuclidCircularB-Regular",
                  "textAlignVertical": "center",
                },
                Array [
                  Object {
                    "color": "#8B939E",
                    "fontSize": 14,
                    "lineHeight": 22,
                  },
                  undefined,
                ],
              ]
            }
          >
            Invalid Date
          </Text>
        </View>
      </View>
      <View
        onLayout={[Function]}
        style={null}
      >
        <View
          style={
            Array [
              Object {
                "flexDirection": "column",
                "width": "100%",
              },
              undefined,
            ]
          }
        >
          <View
            accessible={true}
            focusable={true}
            isTVSelectable={true}
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              Object {
                "alignItems": "center",
                "flexDirection": "row",
                "justifyContent": "center",
                "opacity": 1,
                "paddingBottom": 14,
                "paddingLeft": 20,
                "paddingRight": 20,
                "paddingTop": 14,
                "width": "100%",
              }
            }
          >
            <View
              style={
                Array [
                  Object {
                    "alignItems": "center",
                    "justifyContent": "center",
                    "paddingRight": 12,
                  },
                  undefined,
                ]
              }
            >
              <View
                backgroundColor={null}
                diameter={48}
                style={
                  Array [
                    Object {
                      "alignItems": "center",
                      "backgroundColor": "#EBF0F6",
                      "borderRadius": 24,
                      "height": 48,
                      "justifyContent": "center",
                      "overflow": "hidden",
                      "textAlign": "center",
                      "width": 48,
                    },
                    undefined,
                  ]
                }
              >
                <View
                  name="sent"
                  style={
                    Array [
                      Object {
                        "color": "#BD573A",
                        "fontSize": 48,
                      },
                      undefined,
                    ]
                  }
                />
              </View>
            </View>
            <View
              style={
                Object {
                  "flex": 1,
                }
              }
            >
              <View
                style={
                  Array [
                    Object {
                      "alignItems": "center",
                      "flexDirection": "row",
                      "justifyContent": "space-between",
                      "width": "100%",
                    },
                    undefined,
                  ]
                }
                type="ACTION"
              >
                <View
                  style={
                    Array [
                      Object {
                        "alignItems": "flex-start",
                        "flexBasis": 0,
                        "flexDirection": "column",
                        "flexGrow": 1,
                        "flexShrink": 1,
                        "justifyContent": "center",
                        "marginTop": 0,
                        "minHeight": 54,
                      },
                      Object {
                        "flexGrow": 1,
                      },
                    ]
                  }
                  type="ACTION"
                >
                  <View
                    style={
                      Array [
                        Object {
                          "alignItems": "center",
                          "flexDirection": "row",
                          "justifyContent": "space-between",
                        },
                        undefined,
                      ]
                    }
                  >
                    <Text
                      ellipsizeMode="tail"
                      numberOfLines={2}
                      style={
                        Array [
                          Object {
                            "color": "#0A1427",
                            "fontFamily": "EuclidCircularB-Regular",
                            "textAlignVertical": "center",
                          },
                          Array [
                            Object {
                              "color": "#0A1427",
                              "fontFamily": "EuclidCircularB-Medium",
                              "textAlignVertical": "center",
                            },
                            Array [
                              Object {
                                "color": "#0A1427",
                                "fontSize": 18,
                                "lineHeight": 28,
                                "width": "100%",
                              },
                              undefined,
                            ],
                          ],
                        ]
                      }
                      type="ACTION"
                    >
                      0x0000â€¦000000
                    </Text>
                  </View>
                </View>
                <View
                  rightColumn={true}
                  style={
                    Array [
                      Object {
                        "alignItems": "flex-end",
                        "flexDirection": "column",
                        "justifyContent": "center",
                        "marginLeft": 10,
                        "marginTop": 0,
                        "minHeight": 54,
                      },
                      Object {
                        "maxWidth": "50%",
                      },
                    ]
                  }
                  type="ACTION"
                >
                  <View
                    style={
                      Array [
                        undefined,
                        Object {
                          "flexWrap": "wrap",
                        },
                      ]
                    }
                  >
                    <View
                      horizontal={true}
                      style={
                        Array [
                          Object {
                            "flexDirection": "row",
                            "marginBottom": 0,
                            "marginLeft": 0,
                            "marginRight": 0,
                            "marginTop": 0,
                          },
                          Object {
                            "alignItems": "center",
                            "flexWrap": "wrap",
                            "justifyContent": "flex-end",
                          },
                        ]
                      }
                    >
                      <Text
                        color="#0A1427"
                        ellipsizeMode="tail"
                        numberOfLines={2}
                        style={
                          Array [
                            Object {
                              "color": "#0A1427",
                              "fontFamily": "EuclidCircularB-Regular",
                              "textAlignVertical": "center",
                            },
                            Array [
                              Object {
                                "color": "#0A1427",
                                "fontSize": 18,
                                "lineHeight": 28,
                                "textAlign": "right",
                              },
                              undefined,
                            ],
                          ]
                        }
                      >
                         0 undefined
                      </Text>
                    </View>
                  </View>
                </View>
              </View>
            </View>
          </View>
        </View>
      </View>
    </View>
  </RCTScrollView>
</View>
`;
